pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go

```diff
@@ -137,3 +135,61 @@ func (r *DNSPolicyReconciler) deleteGatewayDNSRecords(ctx context.Context, gatew
 	}
 	return nil
 }
+
+func (r *DNSPolicyReconciler) buildClusterGateway(ctx context.Context, upstreamGateway *gatewayv1beta1.Gateway, downstreamCluster string) (dns.ClusterGateway, error) {
+	var target dns.ClusterGateway
+
+	mc := &clusterv1.ManagedCluster{}
+	if err := r.Client().Get(ctx, client.ObjectKey{Name: downstreamCluster}, mc, &client.GetOptions{}); err != nil {
+		return target, err
+	}
+
+	for _, address := range upstreamGateway.Status.Addresses {
+		if strings.Contains(address.Value, downstreamCluster) {
+			var gatewayAddresses []gatewayv1beta1.GatewayAddress
+
+			addressType := gatewayv1beta1.IPAddressType
+			if strings.Contains(string(*address.Type), string(gateway.MultiClusterHostnameAddressType)) {
+				addressType = gatewayv1beta1.HostnameAddressType
+			}
+
+			tmp := strings.Split(address.Value, "/")
+			addressValue := tmp[0]
+
+			if len(tmp) > 0 {
+				addressValue = tmp[len(tmp)-1]
+			}
+			gatewayAddresses = append(gatewayAddresses, gatewayv1beta1.GatewayAddress{
+				Type:  &addressType,
+				Value: addressValue,
+			})
+			target = *dns.NewClusterGateway(mc, gatewayAddresses)
+		}
+	}
+	return target, nil
+}
+
+func getGatewayAddresses(upstreamGateway *gatewayv1beta1.Gateway) []string {
+	var clusters []string
+
+	for _, address := range upstreamGateway.Status.Addresses {
+		value := strings.Split(address.Value, "/")
+		if strings.Contains(string(*address.Type), string(gateway.MultiClusterHostnameAddressType)) || strings.Contains(string(*address.Type), string(gateway.MultiClusterIPAddressType)) {
+			clusters = append(clusters, value[0])
+		}
+	}
+
+	return clusters
+}
+
+func listenerTotalAttachedRoutes(upstreamGateway *gatewayv1beta1.Gateway, downstreamCluster string) int {
+	listeners := 0
+
+	for _, listener := range upstreamGateway.Status.Listeners {
+		if strings.Contains(string(listener.Name), downstreamCluster) {
+			listeners = int(listener.AttachedRoutes)
+		}
+	}
+
+	return listeners
+}
```

Comment:
This is always going to return the number of attached routes for that last listener. You need to pass in the listener you want the attached routes for and find that one in the list.

I also wouldn't use `strings.Contains`, if someone creates a listener with the same name as a cluster it will give a false positive. You also can't do this for normal gateway listeners, it will only work for MGC ones. I'd split the listener name by '.' instead and assume if you have a length greater than 1 that the cluster name is the first element see https://github.com/Kuadrant/multicluster-gateway-controller/pull/315#discussion_r1259370454.

Reply:
refactored